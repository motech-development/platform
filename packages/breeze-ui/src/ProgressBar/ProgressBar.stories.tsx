import { number, withKnobs } from '@storybook/addon-knobs';
import BaseStyles from '../BaseStyles/BaseStyles';
import ProgressBar from './ProgressBar';

export default {
  component: ProgressBar,
  decorators: [withKnobs],
};

export const BasicProgressBar = {
  name: 'Basic progress bar',
  render: () => (
    <>
      <BaseStyles />

      <ProgressBar max={number('Max', 100)} progress={number('Progress', 50)} />
    </>
  ),
};
