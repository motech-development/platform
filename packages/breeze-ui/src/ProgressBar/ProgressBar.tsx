import { rgba } from 'polished';
import { FC, memo } from 'react';
import styled from 'styled-components';

const ProgressBarInternal = styled.progress`
  appearance: none;
  background-color: ${rgba('#007fa8', 0.25)};
  border: none;
  height: 20px;
  margin: 0;
  padding: 0;
  width: 100%;

  &::-webkit-progress-bar {
    background-color: ${rgba('#007fa8', 0.25)};
    height: 20px;
  }

  &::-webkit-progress-value {
    background-color: #007fa8;
    height: 20px;
  }

  &::-moz-progress-bar {
    background-color: #007fa8;
    height: 20px;
  }
`;

export interface IProgressBarProps {
  max?: number;
  progress: number;
}

const ProgressBar: FC<IProgressBarProps> = ({ max = 100, progress }) => (
  <ProgressBarInternal max={max} value={progress} />
);

export default memo(ProgressBar);
