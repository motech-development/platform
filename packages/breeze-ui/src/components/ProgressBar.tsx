import { forwardRef, ProgressHTMLAttributes } from 'react';
import { classNames, themeClass } from '../utils/className';
import TTheme from '../utils/theme';

export interface IProgressBarProps
  extends ProgressHTMLAttributes<HTMLProgressElement> {
  max?: number;
  progress: number;
  theme?: TTheme;
}

const ProgressBar = forwardRef<HTMLProgressElement, IProgressBarProps>(
  (
    { className = '', max = 100, progress, theme = 'primary', ...rest },
    ref,
  ) => (
    <>
      {/* @tailwind: webkit-progress-value:bg-blue-600 webkit-progress-value:bg-gray-600 webkit-progress-value:bg-green-600 webkit-progress-value:bg-red-600 webkit-progress-value:bg-yellow-600 */}
      {/* @tailwind: webkit-progress-bar:bg-blue-200 webkit-progress-bar:bg-gray-200 webkit-progress-bar:bg-green-200 webkit-progress-bar:bg-red-200 webkit-progress-bar:bg-yellow-200 */}
      {/* @tailwind: moz-progress-bar:bg-blue-600 moz-progress-bar:bg-gray-600 moz-progress-bar:bg-green-600 moz-progress-bar:bg-red-600 moz-progress-bar:bg-yellow-600 */}
      <progress
        className={classNames(
          'appearance-none border-0 p-0 m-0 w-full h-4 webkit-progress-value:h-4 webkit-progress-bar:h-4 moz-progress-bar:h-4',
          themeClass(
            theme,
            'webkit-progress-value:bg-{theme}-600 webkit-progress-bar:bg-{theme}-200 moz-progress-bar:bg-{theme}-600',
          ),
          className,
        )}
        max={max}
        value={progress}
        ref={ref}
        {...rest}
      />
    </>
  ),
);

export default ProgressBar;
