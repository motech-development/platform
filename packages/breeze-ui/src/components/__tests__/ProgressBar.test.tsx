import { render } from '@testing-library/react';
import ProgressBar from '../ProgressBar';

describe('ProgressBar', () => {
  describe('primary', () => {
    it('should render the progress bar without a max value', async () => {
      const { container } = render(<ProgressBar progress={50} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render the progress bar with a max value', async () => {
      const { container } = render(
        <ProgressBar max={80} progress={40} theme="primary" />,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('secondary', () => {
    it('should render the progress bar without a max value', async () => {
      const { container } = render(
        <ProgressBar progress={50} theme="secondary" />,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render the progress bar with a max value', async () => {
      const { container } = render(
        <ProgressBar max={80} progress={40} theme="secondary" />,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('success', () => {
    it('should render the progress bar without a max value', async () => {
      const { container } = render(
        <ProgressBar progress={50} theme="success" />,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render the progress bar with a max value', async () => {
      const { container } = render(
        <ProgressBar max={80} progress={40} theme="success" />,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('warning', () => {
    it('should render the progress bar without a max value', async () => {
      const { container } = render(
        <ProgressBar progress={50} theme="warning" />,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render the progress bar with a max value', async () => {
      const { container } = render(
        <ProgressBar max={80} progress={40} theme="warning" />,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('danger', () => {
    it('should render the progress bar without a max value', async () => {
      const { container } = render(
        <ProgressBar progress={50} theme="danger" />,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render the progress bar with a max value', async () => {
      const { container } = render(
        <ProgressBar max={80} progress={40} theme="danger" />,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
