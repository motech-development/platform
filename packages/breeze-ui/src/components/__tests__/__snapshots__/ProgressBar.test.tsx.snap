// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar danger should render the progress bar with a max value 1`] = `
<progress
  class="appearance-none block border-0 p-0 m-0 w-full h-4 webkit-progress-value:h-4 webkit-progress-bar:h-4 moz-progress-bar:h-4 webkit-progress-value:bg-red-600 webkit-progress-bar:bg-red-200 moz-progress-bar:bg-red-600"
  max="80"
  value="40"
/>
`;

exports[`ProgressBar danger should render the progress bar without a max value 1`] = `
<progress
  class="appearance-none block border-0 p-0 m-0 w-full h-4 webkit-progress-value:h-4 webkit-progress-bar:h-4 moz-progress-bar:h-4 webkit-progress-value:bg-red-600 webkit-progress-bar:bg-red-200 moz-progress-bar:bg-red-600"
  max="100"
  value="50"
/>
`;

exports[`ProgressBar primary should render the progress bar with a max value 1`] = `
<progress
  class="appearance-none block border-0 p-0 m-0 w-full h-4 webkit-progress-value:h-4 webkit-progress-bar:h-4 moz-progress-bar:h-4 webkit-progress-value:bg-blue-600 webkit-progress-bar:bg-blue-200 moz-progress-bar:bg-blue-600"
  max="80"
  value="40"
/>
`;

exports[`ProgressBar primary should render the progress bar without a max value 1`] = `
<progress
  class="appearance-none block border-0 p-0 m-0 w-full h-4 webkit-progress-value:h-4 webkit-progress-bar:h-4 moz-progress-bar:h-4 webkit-progress-value:bg-blue-600 webkit-progress-bar:bg-blue-200 moz-progress-bar:bg-blue-600"
  max="100"
  value="50"
/>
`;

exports[`ProgressBar secondary should render the progress bar with a max value 1`] = `
<progress
  class="appearance-none block border-0 p-0 m-0 w-full h-4 webkit-progress-value:h-4 webkit-progress-bar:h-4 moz-progress-bar:h-4 webkit-progress-value:bg-gray-600 webkit-progress-bar:bg-gray-200 moz-progress-bar:bg-gray-600"
  max="80"
  value="40"
/>
`;

exports[`ProgressBar secondary should render the progress bar without a max value 1`] = `
<progress
  class="appearance-none block border-0 p-0 m-0 w-full h-4 webkit-progress-value:h-4 webkit-progress-bar:h-4 moz-progress-bar:h-4 webkit-progress-value:bg-gray-600 webkit-progress-bar:bg-gray-200 moz-progress-bar:bg-gray-600"
  max="100"
  value="50"
/>
`;

exports[`ProgressBar success should render the progress bar with a max value 1`] = `
<progress
  class="appearance-none block border-0 p-0 m-0 w-full h-4 webkit-progress-value:h-4 webkit-progress-bar:h-4 moz-progress-bar:h-4 webkit-progress-value:bg-green-600 webkit-progress-bar:bg-green-200 moz-progress-bar:bg-green-600"
  max="80"
  value="40"
/>
`;

exports[`ProgressBar success should render the progress bar without a max value 1`] = `
<progress
  class="appearance-none block border-0 p-0 m-0 w-full h-4 webkit-progress-value:h-4 webkit-progress-bar:h-4 moz-progress-bar:h-4 webkit-progress-value:bg-green-600 webkit-progress-bar:bg-green-200 moz-progress-bar:bg-green-600"
  max="100"
  value="50"
/>
`;

exports[`ProgressBar warning should render the progress bar with a max value 1`] = `
<progress
  class="appearance-none block border-0 p-0 m-0 w-full h-4 webkit-progress-value:h-4 webkit-progress-bar:h-4 moz-progress-bar:h-4 webkit-progress-value:bg-yellow-600 webkit-progress-bar:bg-yellow-200 moz-progress-bar:bg-yellow-600"
  max="80"
  value="40"
/>
`;

exports[`ProgressBar warning should render the progress bar without a max value 1`] = `
<progress
  class="appearance-none block border-0 p-0 m-0 w-full h-4 webkit-progress-value:h-4 webkit-progress-bar:h-4 moz-progress-bar:h-4 webkit-progress-value:bg-yellow-600 webkit-progress-bar:bg-yellow-200 moz-progress-bar:bg-yellow-600"
  max="100"
  value="50"
/>
`;
